Source/Core/Core/AchievementManager.cpp

```diff
@@ -204,6 +204,65 @@ void AchievementManager::ActivateDeactivateRichPresence()
       nullptr, 0);
 }
 
+void AchievementManager::DoFrame()
+{
+  if (!m_is_game_loaded)
+    return;
+  rc_runtime_do_frame(
+      &m_runtime,
+      [](const rc_runtime_event_t* runtime_event) {
+        AchievementManager::GetInstance()->AchievementEventHandler(runtime_event);
+      },
+      [](unsigned address, unsigned num_bytes, void* ud) {
+        return AchievementManager::GetInstance()->MemoryPeeker(address, num_bytes, ud);
+      },
+      nullptr, nullptr);
+  u64 current_time = Core::System::GetInstance().GetCoreTiming().GetTicks();
+  if (current_time - m_last_ping_time > SystemTimers::GetTicksPerSecond() * 120)
+    m_queue.EmplaceItem([this] { PingRichPresence(GenerateRichPresence()); });
+}
+
+u32 AchievementManager::MemoryPeeker(u32 address, u32 num_bytes, void* ud)
+{
+  const rc_memory_regions_t* regions = rc_console_memory_regions(m_game_data.console_id);
+  for (u32 ix = 0; ix < regions->num_regions; ix++)
+  {
+    if (address >= regions->region[ix].start_address && address <= regions->region[ix].end_address)
+    {
+      address += (regions->region[ix].real_address - regions->region[ix].start_address);
+      break;
+    }
+  }
+  Core::CPUThreadGuard guard(Core::System::GetInstance());
```

Comment:
Would it be practical to pass the guard as a parameter to `MemoryPeeker` (and maybe `system` as well)? The benefit is that creating `guard` multiple times would pause and then unpause repeatedly (if not on the CPU thread, which I'm not sure is the case here).

Reply:
I actually forgot to remove the guard - I was using it back when the memory accesses were via the MMU. Should I still integrate it here somehow, even if I don't need it for the reads themselves?

Reply:
It depends on if you're using HostRead or Read (and if you know already know you're on the CPU thread or not).

Reply:
Currently this is on the CPU thread. Will using the HookableEvent change threads?